File: python/ex1/gradientDescent.py
import numpy as np

from computeCost import compute_cost
"""梯度下降法"""
def gradient_descent(X,Y,theta_init,alpha,iter_num):
	# 样本个数
	m = Y.shape[0]
	# 代价的历史值
	J_history = np.zeros(iter_num)
	theta = theta_init
	# 进行迭代计算
	for num in range(0,iter_num):
		# 计算每一个theta值下的代价值
		J_history[num] = compute_cost(X,Y,theta)
		# 根据公式计算梯度，来更新theta的值
		hyp = np.dot(X,np.transpose(theta))
		theta = theta - alpha * np.dot(np.transpose(hyp -Y),X) / m

	return theta,J_history

File: matlab/ex1/gradientDescent.m
function [theta_Final , thetaAll] = gradientDescent ( X,Y,theta_init,alpha,num_itera )
%% 函数功能：梯度下降法求解theta值
%  X：输入
%  Y：对应输出
%  alpha：学习率/步长
%  theta_Init：theta初始值
%  num_itera：迭代计算次数
%%
[n,m]=size(X);
theta = theta_init;

thetaAll = zeros(n,num_itera);
thetaAll(:,1) = theta_init;

J = zeros(num_itera,1);
% 保存每一次迭代的代价值，并更新theta
for i = 1:num_itera
    J(i) = computeCost(X,Y,theta);
    theta = theta - alpha * (1/m) * (X * (theta.' * X - Y).');
    thetaAll(:,i) = theta;
end
theta_Final = theta;
end

